Sources/GRPC/AsyncAwaitSupport/GRPCAsyncServerHandler.swift

```diff
@@ -0,0 +1,632 @@
+        /// receiving `.end`. Therefore this `.message` must have been sent by the client after it
+        /// sent `.end`, which is a protocol violation.
+        self.handleError(GRPCError.ProtocolViolation("Message received after end of stream"))
+      }
+    case .completed:
+      // We received a message but we're already done: this may happen if we terminate the RPC
+      // due to a channel error, for example.
+      ()
+    }
+  }
+
+  @inlinable
+  internal func receiveInterceptedEnd() {
+    switch self.state {
+    case .idle:
+      self.handleError(GRPCError.ProtocolViolation("End of stream received before headers"))
+    case let .active(requestStreamSource, _, _, _):
+      switch requestStreamSource.finish() {
+      case .accepted(queueDepth: _):
+        break
+      case .dropped:
+        /// If we are in the `.active` state then we have yet to encounter an error. Therefore
+        /// if the request stream source has already terminated then it must have been the result of
+        /// receiving `.end`. Therefore this `.end` must have been sent by the client after it
+        /// already sent `.end`, which is a protocol violation.
+        self.handleError(GRPCError.ProtocolViolation("Message duplicate end of stream"))
+      }
+    case .completed:
+      // We received a message but we're already done: this may happen if we terminate the RPC
+      // due to a channel error, for example.
+      ()
+    }
+  }
+
+  // MARK: - User Function To Interceptors
+
+  @inlinable
+  internal func interceptResponse(
+    _ response: Response,
+    metadata: MessageMetadata,
+    promise: EventLoopPromise<Void>?
+  ) {
+    self.context.eventLoop.assertInEventLoop()
+    switch self.state {
+    case .idle:
+      // The user handler cannot send responses before it has been invoked.
+      preconditionFailure()
+
+    case .active:
+      self.interceptors.send(.message(response, metadata), promise: promise)
+
+    case .completed:
+      promise?.fail(GRPCError.AlreadyComplete())
+    }
+  }
+
+  @inlinable
+  internal func userHandlerCompleted(_ result: Result<Void, Error>) {
+    switch self.state {
+    case .idle:
+      // The user handler cannot complete before it is invoked.
+      preconditionFailure()
+
+    case .active:
+      switch result {
+      case .success:
+        /// The user handler has completed successfully.
+        /// We don't take any action here; the state transition and termination of the message
+        /// stream happen when the response stream has drained, in the response stream writer
+        /// delegate callback, `responseStreamDrained(_:)`.
+        break
+
+      case let .failure(error):
+        self.handleError(error, thrownFromHandler: true)
+      }
+
+    case .completed:
+      ()
+    }
+  }
+
+  @inlinable
+  internal func responseStreamDrained(_ status: GRPCStatus) {
+    switch self.state {
```

Comment:
Let's assert we're on the correct event loop in here

Reply:
Done when adding the EL dance in 0c819a55